platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/AsyncPreConsensusEventWriter.java

```diff
@@ -0,0 +1,176 @@
+    @Override
+    public void start() {
+        writer.start();
+        handleThread.start();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void stop() {
+        handleThread.stop();
+        writer.stop();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void writeEvent(final EventImpl event) throws InterruptedException {
+        if (event.getStreamSequenceNumber() == EventImpl.NO_STREAM_SEQUENCE_NUMBER
+                || event.getStreamSequenceNumber() == EventImpl.STALE_EVENT_STREAM_SEQUENCE_NUMBER) {
+            throw new IllegalStateException("Event must have a valid stream sequence number");
+        }
+        eventInserter.put(event);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setMinimumGenerationNonAncient(final long minimumGenerationNonAncient) throws InterruptedException {
+        minimumGenerationNonAncientInserter.put(minimumGenerationNonAncient);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setMinimumGenerationToStore(final long minimumGenerationToStore) {
+        writer.setMinimumGenerationToStore(minimumGenerationToStore);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isEventDurable(final EventImpl event) {
+        return writer.isEventDurable(event);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void waitUntilDurable(final EventImpl event) throws InterruptedException {
+        writer.waitUntilDurable(event);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean waitUntilDurable(final EventImpl event, final Duration timeToWait) throws InterruptedException {
+        return writer.waitUntilDurable(event, timeToWait);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void requestFlush(EventImpl event) {
+        writer.requestFlush(event);
+    }
+
+    /**
+     * Pass a minimum generation non-ancient to the wrapped writer.
+     */
+    private void setMinimumGenerationNonAncientHandler(final Long minimumGenerationNonAncient) {
+        // Unless we do something silly like wrapping an asynchronous writer inside another asynchronous writer,
+        // this should never throw an InterruptedException.
+        abortAndThrowIfInterrupted(
+                () -> writer.setMinimumGenerationNonAncient(minimumGenerationNonAncient),
+                "interrupted while attempting to call setMinimumGenerationNonAncient on writer");
+    }
+
+    /**
+     * Pass an event to the wrapped writer.
+     */
+    private void addEventHandler(final EventImpl event) {
+        // Unless we do something silly like wrapping an asynchronous writer inside another asynchronous writer,
+        // this should never throw an InterruptedException.
+        abortAndThrowIfInterrupted(
+                () -> writer.writeEvent(event), "interrupted while attempting to call addEvent on writer");
```

Comment:
I don't think creating a lambda for every event is wise, can we avoid this?

Reply:
refactored to avoid lambdas